### PR TITLE
Split Unit duties into a Hangar class

### PR DESCRIPTION
**Pull request by sixlettervariables**
This PR is part of a project of mine to reduce the cognitive burden of Campaign.java and move it into smaller, easier to understand classes. The first step I took was to move the "accounting" functions of Units into a logical Hangar class. This _could_ mean a foray into tracking multiple units on multiple places...but let's not go there just yet.

This also significantly reduces the memory burden of certain areas and on larger campaigns certain actions will be faster, as there won't be giant copies of the unit list made.

**Note: this is a draft while I do some more testing, but functionally this was a lift-n-shift.**

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 1]
```diff
@@ -3676,24 +3617,19 @@ private Money getTheoreticalPayroll(boolean noInfantry) {
...
-        Money costs = Money.zero();
         if (campaignOptions.payForMaintain()) {
-            for (Map.Entry<UUID, Unit> mu : units.entrySet()) {
-                Unit u = mu.getValue();
-                if (u.requiresMaintenance() && null != u.getTech()) {
-                    costs = costs.plus(u.getMaintenanceCost());
-                }
-            }
+            return getHangar().getUnitsStream()
+                .filter(u -> u.requiresMaintenance() && null != u.getTech())
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:3622]:
```suggestion
                .filter(u -> u.requiresMaintenance() && (null != u.getTech()))
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 2]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
     }
 
     public int getTotalMechBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getMechCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7175]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 3]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalASFBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getASFCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7181]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 4]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalBattleArmorBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getBattleArmorCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7193]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 5]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalSmallCraftBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getSmallCraftCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7187]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 6]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalLightVehicleBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getLightVehicleCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7211]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 7]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalHeavyVehicleBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getHeavyVehicleCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7205]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 8]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalInfantryBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getInfantryCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7199]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 9]
```diff
@@ -7236,117 +7172,88 @@ public void setHealingTimeOptions(int newHeal, int newNaturalHeal) {
...
+            .sum());
     }
 
     public int getTotalProtomechBays() {
-        double bays = 0;
-        for (Unit u : getUnits()) {
-            bays += u.getProtomechCapacity();
-        }
-        return (int)Math.round(bays);
+        return (int)Math.round(getHangar().getUnitsStream()
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:7217]:
```suggestion
        return (int) Math.round(getHangar().getUnitsStream()
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 10]
```diff
@@ -8194,8 +8101,8 @@ public void initAtB(boolean newCampaign) {
                                     // TODO : Fix this so we aren't using a hack that just assumes IS2
                                     p.setOriginalUnitTech(Person.TECH_IS2);
                                 }
-                                if (null != p.getUnitId() && null != units.get(p.getUnitId())
-                                        && ms.getName().equals(units.get(p.getUnitId()).getEntity().getShortNameRaw())) {
+                                if (null != p.getUnitId() && null != getHangar().getUnit(p.getUnitId())
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:8104]:
```suggestion
                                if ((null != p.getUnitId()) && (null != getHangar().getUnit(p.getUnitId()))
```

**Review by Windchild292 — APPROVED** [review]
Changes look good, just a few formatting and other comments

File: MekHQ/src/mekhq/gui/dialog/PersonnelMarketDialog.java [hunk 11]
```diff
@@ -365,14 +365,7 @@ private void addUnit(Entity en, boolean pay) {
...
-		for (Unit u : campaign.getUnits()) {
-			if (u.getEntity() == en) {
-				unit = u;
-				break;
-			}
-        }
+		Unit unit = campaign.addNewUnit(en, false, 0);
         if (unit == null) {
             // No such unit matching the entity.
             return;
```
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/gui/dialog/PersonnelMarketDialog.java:371]: Swore I got all of these, may not have pushed the change.

**Pull request merged**